server/updater/updater.go

```diff
@@ -0,0 +1,295 @@
+	}
+}
+
+// NotificationsOption sets a notification chan to receive update related notifications.
+// these notifications sent for every successful and unsuccessful updates and any errors
+// occurred during an update process.
+// notifications needs to be consumed within separate goroutines in order to not block the updater.
+// notifications chan cannot be updated by the UpdateConfig(). trying it has no effects.
+// notifications chan will be closed automatically after Updater is stopped.
+func NotificationsOption(notifications chan Notification) Option {
+	return func(u *Updater) {
+		u.conf.notifications = notifications
+	}
+}
+
+// Start starts updater to regularly check and do plugin updates.
+// Start() blocks until pooling is cancelled and if there is any, waits for current update
+// proccess to be competed before returning.
+// calling Start() multiple times is a misuse. caller should initialize and use a new Updater.
+func (u *Updater) Start() error {
+	u.stopWait.Add(1)
+	defer func() {
+		defer u.stopWait.Done()
+		if u.notifications != nil {
+			close(u.notifications)
+		}
+	}()
+	// create a cancelable context to stop pooling later.
+	var ctx context.Context
+	ctx, u.stopPooling = context.WithCancel(context.Background())
+	// use a distributed lock here to only check and do updates in a single node(plugin instance)
+	// at the same time.
+	dl := dlock.New(updateLockKey, u.dlockStore)
+	for {
+		if err := dl.Lock(dlock.ContextOption(ctx)); err != nil {
+			if err == context.Canceled {
+				return nil
+			}
+			return err
+		}
+		// check and do updates.
+		u.checkAndUpdate()
+		// if there is a stop signal, it has the priority.
+		select {
+		case <-ctx.Done():
+			dl.Unlock()
+			return nil
+		default:
+		}
+		conf := u.cloneConfing()
+		afterC := time.After(conf.updateInterval)
+		// wait for the next update round or a stop signal.
+		select {
+		// wait for updateInterval before doing the next check.
+		// we start waiting after running the update check to make sure that once the Updater is
+		// started, it will immediately start to the first update process instead of waiting.
+		case <-afterC:
+			dl.Unlock()
+		case <-ctx.Done():
+			dl.Unlock()
+			return nil
+		}
+	}
+}
+
+// checkAndUpdate checks for new versions of installed plugins and updates them accordingly.
+func (u *Updater) checkAndUpdate() {
+	u.papi.LogInfo("checking for new versions...")
+	updates := u.discover()
+	lenUpdates := len(updates)
+	if lenUpdates == 0 {
+		u.papi.LogInfo("no new versions found")
+		return
+	}
+	u.papi.LogInfo(fmt.Sprintf("found %d plugins to update", lenUpdates))
+	var wg sync.WaitGroup
+	wg.Add(lenUpdates)
+	// TODO(ilgooz): limit the number of how many goroutines can be created.
+	for _, updateOp := range updates {
+		go func() {
+			defer wg.Done()
+			u.papi.LogInfo(fmt.Sprintf("updating %q from %q to %q...", updateOp.installed.Id,
+				updateOp.installed.Version, updateOp.next.Manifest.Version))
+			u.update(updateOp)
```

Comment:
if `update()` is unsuccessful we cannot print the log message in the following line.